### PR TITLE
[iOS] - Add & Export iOS API to use AIChat

### PR DESCRIPTION
**Pull request by Brandon-T**
## Security Review
- https://github.com/brave/reviews/issues/1463

## Summary
- Export iOS API to use AIChat
- Fix framework versioning bug to follow Desktop

Resolves https://github.com/brave/brave-ios/issues/8125

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/browser/api/ai_chat/ai_chat_conversation_driver.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver.h:57]: Missing `__weak` here?

File: ios/browser/api/ai_chat/ai_chat_factory+private.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat_factory+private.mm:1]: This file doesn't need `+private` in the name

File: ios/browser/api/ai_chat/ai_chat_factory+private.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat_factory+private.mm:14]: __weak here is redundant and doesn't affect anything

File: ios/browser/api/ai_chat/ai_chat_factory.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat_factory.h:18]: __weak here is redundant and doesn't affect anything

File: ios/app/brave_core_main.mm
Comment (kylehickinson) [inline ios/app/brave_core_main.mm:354]: The factory class here seems a bit pointless? Why not just make a method here that returns an AIChat instance directly and completely omit the factory?

File: ios/browser/api/BUILD.gn
Comment (kylehickinson) [inline ios/browser/api/BUILD.gn:9]: Don't think `component` is the right template to use here? And if the deps are listed here then `ios/browser` likely doesn't need. any of the `api/*` deps listed in it anymore and would just need `//brave/ios/browser/api` (except for any apis that are directly used in `brave_web_client`/`brave_web_main_parts` but I dont think there are any)

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:228]: Is this logic not shared with desktop?

File: ios/browser/api/ai_chat/ai_chat.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.h:23]: Remove custom getters

File: ios/browser/api/ai_chat/ai_chat.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.h:38]: Remove custom getter

File: ios/browser/api/ai_chat/ai_chat.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.h:30]: Remove custom getter

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:41]: __weak here is redundant and doesnt affect anything

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:55]: Probably better to just get the pref service and store it for the lifetime of the instance, no?

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:69]: Seems like this sort of logic should be done in `components/ai_chat` not iOS specific

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:90]: Return `[models copy]` to ensure an immutable array is returned

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.mm:99]: Return `[array copy]` to ensure an immutable array is returned

**Review by kylehickinson — CHANGES_REQUESTED** [review]
I see a lot of duplicated logic that exists in https://github.com/brave/brave-core/blob/ios-creds-provider/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

Is there no way to shift that logic into something shared by all platforms?

**Comment by petemill** [comment]
> I see a lot of duplicated logic that exists in https://github.com/brave/brave-core/blob/ios-creds-provider/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc
> 
> Is there no way to shift that logic into something shared by all platforms?

Yes I've suggested we convert to a ViewModel-like structure in C++, and include more of the UI state handling that we have in the class you're pointing at and in https://github.com/brave/brave-core/blob/master/components/ai_chat/resources/page/state/data-context-provider.tsx, that iOS and brave-core could both use.

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: ios/browser/api/ai_chat/ai_chat.h
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat.h:55]: Missing `_Nullable` on the identifier argument (method impl can call `completion` with `nil`)

File: ios/browser/api/ai_chat/ai_chat_conversation_driver.mm
Comment (kylehickinson) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver.mm:49]: Small nit: `net::GURLWithNSURL` already does this nullability check, so this can all be simplified to 

```objc
return net::GURLWithNSURL([bridge_ getLastCommittedURL])
```
or 
```objc
NSURL* url = [bridge_ getLastCommittedURL];
return net::GURLWithNSURL(url)
```
if you prefer to keep the local var

**Review by kylehickinson — CHANGES_REQUESTED** [review]

**Review by kylehickinson — APPROVED** [review]

File: ios/app/BUILD.gn
Comment (bridiver) [inline ios/app/BUILD.gn:1]: copyright date shouldn't change

Comment (Brandon-T) [reply ios/app/BUILD.gn]: Reverted.

File: ios/browser/api/ai_chat/ai_chat_conversation_driver.mm
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver.mm:80]: passing in the browser state and using it to get other services from the factory is an anti-pattern in chromium. You should pass in the skus service, prefs and shared url loader factory into the `AIChatConversationDriverIOS`, but I'm concerned about lifetimes here, is an `AIChatConversationDriver` explicitly tied to a tab in ios and are all tabs destroyed before we shutdown the chromium services?

File: ios/browser/api/ai_chat/ai_chat_delegate.h
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat_delegate.h:18]: why is this method named `WithCompletion`? We generally try to keep the same method names across chromium/ios

File: ios/browser/api/ai_chat/ai_chat_conversation_driver.mm
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver.mm:6]: filename should match the classname ai_chat_conversation_driver_ios.mm

File: ios/browser/api/ai_chat/ai_chat_delegate.h
Comment (Brandon-T) [reply ios/browser/api/ai_chat/ai_chat_delegate.h:18]: This is an iOS thing: https://developer.apple.com/documentation/swift/calling-objective-c-apis-asynchronously and doesn't have anything to do with Chromium.

Chromium also does the same thing: https://source.chromium.org/chromium/chromium/src/+/main:ios/chrome/browser/ui/authentication/authentication_flow.mm;l=170

File: ios/browser/api/ai_chat/ai_chat_conversation_driver.mm
Comment (Brandon-T) [reply ios/browser/api/ai_chat/ai_chat_conversation_driver.mm]: Fixed. Renamed the class. VPN and all classes that use Skus pass in a getter. Refactoring it would mean I have to refactor Desktop and Android again.

File: components/ai_chat/core/browser/conversation_driver.cc
Comment (bridiver) [inline components/ai_chat/core/browser/conversation_driver.cc:50]: per DM I don't think these are the right changes to make because they require a subclass of ConversationDriver on the ios side. Instead you should keep the url loader factory as a param and also pass in the things you need to create AIChatCredentialManager and AIChatFeedbackAPI, not the classes themselves

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat.mm:26]: is this used here?

File: ios/app/brave_main_delegate.mm
Comment (bridiver) [inline ios/app/brave_main_delegate.mm:75]: is this a rebase issue? It doesn't seem related to this PR?

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (bridiver) [inline components/ai_chat/content/browser/ai_chat_tab_helper.cc:20]: this is a layering violation, you need to pass the channel into the TabHelper

File: components/ai_chat/content/browser/DEPS
Comment (bridiver) [inline components/ai_chat/content/browser/DEPS:2]: see https://github.com/brave/brave-core/pull/21032/files#r1417712279

File: chromium_src/ios/chrome/browser/shared/model/prefs/BUILD.gn
Comment (bridiver) [inline chromium_src/ios/chrome/browser/shared/model/prefs/BUILD.gn:7]: also a rebase issue maybe? This doesn't seem related to this PR

File: ios/browser/api/ai_chat/DEPS
Comment (bridiver) [inline ios/browser/api/ai_chat/DEPS:4]: you might want to configure your editor to automatically add a newline at the end so you don't get these

File: components/ai_chat/core/browser/conversation_driver.cc
Comment (Brandon-T) [reply components/ai_chat/core/browser/conversation_driver.cc]: Fixed

File: ios/app/brave_main_delegate.mm
Comment (Brandon-T) [reply ios/app/brave_main_delegate.mm:75]: It's not related, but it fixes griffin on iOS and seemed like a small change. It's related to: https://bravesoftware.slack.com/archives/C7VLGSR55/p1700081786169829

I just put it in this PR instead of a separate one.

File: components/ai_chat/content/browser/ai_chat_tab_helper.cc
Comment (Brandon-T) [reply components/ai_chat/content/browser/ai_chat_tab_helper.cc]: Fixed.

File: chromium_src/ios/chrome/browser/shared/model/prefs/BUILD.gn
Comment (Brandon-T) [reply chromium_src/ios/chrome/browser/shared/model/prefs/BUILD.gn]: Fixed

File: ios/app/brave_main_delegate.mm
Comment (bridiver) [reply ios/app/brave_main_delegate.mm:75]: It's generally better to get unrelated issues separate for dealing with rollbacks, uplifts, debugging, etc... so I think it would be better as a separate PR

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (Brandon-T) [reply ios/browser/api/ai_chat/ai_chat.mm]: Removed.

File: ios/browser/api/ai_chat/ai_chat.mm
Comment (Brandon-T) [reply ios/browser/api/ai_chat/ai_chat.mm]: Done.

File: ios/app/brave_main_delegate.mm
Comment (Brandon-T) [reply ios/app/brave_main_delegate.mm]: Done

File: ios/browser/api/ai_chat/ai_chat_conversation_driver_ios.mm
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver_ios.mm:7]: this should be in sorted order below with the other includes

File: ios/browser/api/ai_chat/ai_chat_conversation_driver_ios.mm
Comment (bridiver) [inline ios/browser/api/ai_chat/ai_chat_conversation_driver_ios.mm:23]: `ConversationDriverIOS`

File: ios/browser/api/ai_chat/BUILD.gn
Comment (bridiver) [inline ios/browser/api/ai_chat/BUILD.gn:60]: nit: this should just be `mojom`

**Review by bridiver — APPROVED** [review]
approving assuming the remaining small changes will be made

**Comment by kylehickinson** [comment]
Since you have changes outside of the ios directly I removed the CI skip labels, please ensure your PR is built on all CI targets before merging

**Pull request merged**
